toolchain/semantics/semantics_handle.cpp

```diff
@@ -43,13 +43,44 @@ auto SemanticsHandleDeducedParameterListStart(SemanticsContext& context,
   return context.TODO(parse_node, "HandleDeducedParameterListStart");
 }
 
-auto SemanticsHandleDesignatorExpression(SemanticsContext& context,
+auto SemanticsHandleQualifiedDeclaration(SemanticsContext& context,
                                          ParseTree::Node parse_node) -> bool {
+  // The child parse nodes are in reverse order versus how they need to be
+  // applied.
+  auto [parse_node2, node_or_name_id2] =
+      context.node_stack().PopWithParseNode<SemanticsNodeId>();
+  if (context.parse_tree().node_kind(context.node_stack().PeekParseNode()) ==
+      ParseNodeKind::QualifiedDeclaration) {
+    // QualifiedDeclarations have already been applied.
+    context.node_stack().PopAndDiscardSoloParseNode(
+        ParseNodeKind::QualifiedDeclaration);
+  } else {
+    auto [parse_node1, node_or_name_id1] =
```

Comment:
Can we `CHECK` that the top of the declaration name stack is in the `New` state here? I think that'd make the behavior here a little easier for me to follow.

Reply:
This is a little awkward -- we'd need to either add a function or a parameter for a check, since the stack is an implementation detail of the context. Do you still want it?

I've added a comment to clarify behavior. Also, shifted the behavior a little to reduce pushes/pops.

Reply:
Let's leave it as-is; I think the added comments help sufficiently.